Add Expected Attention with Stats

alessiodevoto opened the pull request:
## PR description

So far Expected Attention only supports non QK norm models (no support for Gemma and Qwen) and it requires computing mean and covariance at inference time. This PR adds the option to load precomputed stats if they are available, or compute them on the fly if not. There are two main contributions:

- Expected attention with stats
- Change to the base press to have it work with subpresses that require __post_init_from_model__.

## Checklist

- Tests are working (`make test`)
- Code is formatted correctly (`make style`, on errors try fix with `make format`)
- Copyright header is included
- [x] All commits are signed-off  using `git commit -s`
- [ ] (new press) `mypress_press.py` is in the `presses` directory
- [ ] (new press) `MyPress` is in `__init__.py` 
- [ ] (new press) `README.md` is updated with a 1 liner about the new press in the Available presses section
- [ ] (new press) New press is in the `default_presses` list in `tests/default_presses.py`
- [ ] (new press) A docstring is provided that follows the same structure as the existing ones

alessiodevoto commented on kvpress/presses/expected_attention_press.py:
```diff
@@ -166,3 +227,38 @@ def score(
+                self.mu = self.stats.query_mean
+                self.cov = self.stats.query_cov
+            else:
+                print(f"Could not load query statistics for {stats_id} from the HF hub.")
```
It does monkey patching but just as a context manager so all changes are removed right after it is applied. Also, this is coherent with what we do in duo attention (computing stats on the fly) to avoid the cumbersome process of doing the computation offlin. However, I will move it to a separate file, ask the user to run the code, store the results, then rerun the press by passing the saved files.

maxjeblick replied:
> It does monkey patching but just as a context manager so all changes are removed right after it is applied
Yes, that's correct. It is more about code isolation of non-standard python code.

maxjeblick replied:
> Ok, should we also do the same for duo attention then ?

No, I think it's fine to leave for dua attention as is.

To summarize again:
- The code `collect_queries` works and restores the original state
- `collect_queries` uses some advanced import monkey-patch hacks, which are I'd like to keep out of the main kvpress code.

alessiodevoto commented on kvpress/presses/utils.py:
```diff
@@ -0,0 +1,163 @@
+        inputs = tokenizer(text, return_tensors="pt").to(model.device)
+        with patch_rotary_embedding(model) as captured_queries:
+            model(**inputs)
+        collected_queries.append(torch.cat(captured_queries, dim=0)[:, :, n_sink:, :].cpu())
```
Just left it for consistency with the original method, where sinks are not considered for stats

maxjeblick replied:
Ok, makes sense, could you add a comment

alessiodevoto commented on kvpress/presses/utils.py:
```diff
@@ -0,0 +1,163 @@
+
+
+@torch.inference_mode()
+def compute_query_covariance(queries: list[torch.Tensor]) -> torch.Tensor:
```
Ops stayed there from previous version

alessiodevoto commented on kvpress/presses/expected_attention_press.py:
```diff
@@ -60,11 +103,27 @@ class ExpectedAttentionPress(ScorerPress):
         Compute the mean and covariance matrix of the queries
         """
+        if self.use_stats:
+            if self.mu is None or self.cov is None:
```
I think we discussed this with Simon and we decided that it was better not to have a new separate press. I agree as it is the main logic, but open to change

maxjeblick replied:
After some consideration, I think it is better to have a new press:
- 3 new parameters are added which adds complexity to the existing press
- Press signature would change
- More entangled code

maxjeblick commented on kvpress/presses/expected_attention_press.py:
```diff
@@ -60,11 +103,27 @@ class ExpectedAttentionPress(ScorerPress):
+    # These are only used if use_stats is True
+    stats_dataset: str = "kmfoda/booksum"
+    n_samples: int = 100
+    mu: torch.Tensor = None  # populated by __post_init_from_model__ (num_layers, num_heads, head_dim)
```
They are class attributes if they are initialized in post_init

maxjeblick commented on kvpress/presses/expected_attention_press.py:
```diff
@@ -14,6 +18,37 @@
+logger = logging.getLogger(__name__)
+
+
+class ExpectedAttentionStats(torch.nn.Module, PyTorchModelHubMixin):
```
> I left it here to make it consistent with the other classes

It would make `utils.py` self-consistent to compute and store mean and covariance.

alessiodevoto replied:
Ok, should we adjust QFilters accordingly, i.e. move the concept of QFilters which is specific to that press to utils ?

maxjeblick replied:
No, I wouldn't.
utils.py (or `calculate_expected_attention_stats.py`) is a self-contained file that claculates stats.
For other presses, we don't need to move things.

alessiodevoto commented on kvpress/presses/base_press.py:
```diff
@@ -175,6 +179,10 @@ def __call__(self, model: PreTrainedModel) -> Generator:
         ...     # Forward pass with compression applied
         ...     outputs = model(input_ids, past_key_values=cache)
         """
+        if hasattr(self, "press"):
```
Ok, we have two options:
- in all presses that wrap a press, call __post_init_from_model__ ---> a lot of duplicated code, but makes it clear that that press is taking care of it
- have it in base press, less code, more straightwforward 

I tried both option (started from the first then moved to the second as it seemed more elegant). Open to switch back to the previous one. In general, I think this feature is quite important (a lot of presses are wrapped)

alessiodevoto replied:
This will go in a different pr

maxjeblick commented on evaluation/evaluate_config.yaml:
```diff
@@ -1,22 +1,22 @@
 compress_questions: false                          # Whether to compress questions with context
 
-device: null  # Device to use (null = auto-detect, "cuda:0", "cpu", etc.)
+device: "cuda:1"  # Device to use (null = auto-detect, "cuda:0", "cpu", etc.)
```
null = auto-detect, 

can you elave it as is?

alessiodevoto commented:
I have refactored the code significantly (sorry 😅), before computing and uploading stats to the hub and adding the press to tests and press list I would like to hear your opinion :) 

Major changes:
- updated ea code and made it more modular (this way it is easier to subclass)
- instead of having a separate file just for computing stats, I put everything inside the press file. This way it stays self-contained and cleaner. 

Let me know wdyt @maxjeblick 🙂

alessiodevoto commented:
No problem I will update later as well. 
- For fire instead of argparse -> Ok. 
- As for the comments, I think I addressed all of them in this PR. The only difference that I wanted to double check is whether we should have an external file for stats computation or it is better to keep it self contained like for duo attention.

alessiodevoto commented:
Got it, I simplified the if else logic. Also, there was a bug in the computation (I was computing the stats but forgot to save them), now it is fixed.

alessiodevoto commented:
Update: 
- computed stats both for mock testing model and Qwen (for future)
- added tests (they download stats from the hub)

alessiodevoto commented:
/ok to test 7f93961

alessiodevoto commented:
Forgot to clean some comments, now it's should be /ok to test bc27dd8

alessiodevoto commented:
/ok to test bc27dd8
